### PR TITLE
Fix stack overflow in DebugFileIO::RecursiveCreateDir for relative debug dump paths (#123114)

### LINKED ISSUE #123114: SavedModels containing `DebugIdentity/DebugNanCount/DebugNumericSummary` raw ops segfault at inference
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.21.0 and 2.22.0-dev20260706

### Custom code

Yes

### OS platform and distribution

Ubuntu 22.04, Linux 6.8.0-100-generic x86_64

### Mobile device

Ubuntu 22.04, Linux 6.8.0-100-generic x86_64

### Python version

3.11.15

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

CUDA 13.0

### GPU model and memory

_No response_

### Current behavior?

SavedModels that contain TensorFlow debug raw ops can be saved successfully, but
loading and running inference crashes the Python process with SIGSEGV.

I reproduced this with:

- `tf.raw_ops.DebugIdentity`
- `tf.raw_ops.DebugIdentityV3`
- `tf.raw_ops.DebugNanCount`
- `tf.raw_ops.DebugNumericSummary`

This still reproduces with TensorFlow nightly:

```text
tensorflow 2.22.0-dev20260706
git v1.12.1-142040-g37d07506077
```

### Standalone code to reproduce the issue

```python
# Each of the following variants creates a SavedModel that later segfaults during inference. 
# The same `load.py` below can be used for all of them.


# tf.raw_ops.DebugIdentity

import tensorflow as tf


class ReproModel(tf.Module):
    @tf.function(input_signature=[tf.TensorSpec(shape=[None], dtype=tf.float32)])
    def __call__(self, x):
        debug_op = tf.raw_ops.DebugIdentity(
            input=x,
            device_name="/device:CPU:0",
            tensor_name="sample_tensor:0",
            debug_urls=["file://tfdbg_DebugIdentity"],
            gated_grpc=False,
            name="sample_debug_identity",
        )
        zero_impact = tf.reduce_sum(tf.cast(debug_op, tf.float32)) * 0.0
        return x + zero_impact


model = ReproModel()
tf.saved_model.save(model, "model")


# -----------------------------------
# tf.raw_ops.DebugIdentityV3


import tensorflow as tf


class ReproModel(tf.Module):
    @tf.function(input_signature=[tf.TensorSpec(shape=[None], dtype=tf.float32)])
    def __call__(self, x):
        debug_op = tf.raw_ops.DebugIdentityV3(
            input=x,
            device_name="/device:CPU:0",
            tensor_name="sample_tensor:0",
            io_of_node="sample_tensor",
            is_input=False,
            io_index=0,
            debug_urls=["file://tfdbg_DebugIdentityV3"],
            gated_grpc=False,
            name="sample_debug_identity_v3",
        )
        zero_impact = tf.reduce_sum(tf.cast(debug_op, tf.float32)) * 0.0
        return x + zero_impact


model = ReproModel()
tf.saved_model.save(model, "model")


# -----------------------------------
# tf.raw_ops.DebugNanCount

import tensorflow as tf


class ReproModel(tf.Module):
    @tf.function(input_signature=[tf.TensorSpec(shape=[None], dtype=tf.float32)])
    def __call__(self, x):
        debug_op = tf.raw_ops.DebugNanCount(
            input=x,
            device_name="/device:CPU:0",
            tensor_name="sample_tensor:0",
            debug_urls=["file://tfdbg_DebugNanCount"],
            gated_grpc=False,
            name="sample_debug_nan_count",
        )
        zero_impact = tf.reduce_sum(tf.cast(debug_op, tf.float32)) * 0.0
        return x + zero_impact


model = ReproModel()
tf.saved_model.save(model, "model")


# -----------------------------------
# tf.raw_ops.DebugNumericSummary

import tensorflow as tf


class ReproModel(tf.Module):
    @tf.function(input_signature=[tf.TensorSpec(shape=[None], dtype=tf.float32)])
    def __call__(self, x):
        debug_op = tf.raw_ops.DebugNumericSummary(
            input=x,
            device_name="/device:CPU:0",
            tensor_name="sample_tensor:0",
            debug_urls=["file://tfdbg_local_dump"],
            lower_bound=-1.0,
            upper_bound=1.0,
            mute_if_healthy=False,
            gated_grpc=False,
            name="sample_debug_numeric_summary",
        )
        zero_impact = tf.cast(debug_op[0], tf.float32) * 0.0
        return x + zero_impact


model = ReproModel()
tf.saved_model.save(model, "model")


# -----------------------------------
# Shared load.py for all four SavedModels.

import tensorflow as tf

loaded = tf.saved_model.load("model")
print(loaded(tf.constant([10.0, -5.0, 0.0, 8.0], dtype=tf.float32)))
```

### Relevant log output

```shell
Segmentation fault (core dumped)
```

### PR DESCRIPTION
**Pull request by somtri**
Fixes #123114

## Problem

SavedModels containing `tf.raw_ops.DebugIdentity`, `DebugIdentityV3`, `DebugNanCount`, or `DebugNumericSummary` with a relative `file://` debug URL (e.g. `debug_urls=["file://tfdbg_DebugIdentity"]`, no leading slash) segfault when loaded and run in a fresh process:

```python
import tensorflow as tf

class ReproModel(tf.Module):
  @tf.function(input_signature=[tf.TensorSpec(shape=[None], dtype=tf.float32)])
  def __call__(self, x):
    debug_op = tf.raw_ops.DebugIdentity(
        input=x, device_name="/device:CPU:0", tensor_name="sample_tensor:0",
        debug_urls=["file://tfdbg_DebugIdentity"], gated_grpc=False)
    return x + tf.reduce_sum(tf.cast(debug_op, tf.float32)) * 0.0

tf.saved_model.save(ReproModel(), "model")
```

Then, in a separate process:

```python
import tensorflow as tf

m = tf.saved_model.load("model")
print(m(tf.constant([10.0, -5.0, 0.0, 8.0])))
```

```
$ python load_and_run.py
Segmentation fault (core dumped)
```

## Root cause

`DebugFileIO::RecursiveCreateDir` (`tensorflow/core/debug/debug_io_utils.cc`) walks up a directory path calling `io::Dirname()` to find each parent, recursing until it finds one that already exists. For a relative, slash-less path (like `tfdbg_DebugIdentity`), `io::Dirname()` returns an empty string, and `io::Dirname("")` returns `""` again: a fixed point with no base case to stop the recursion. `RecursiveCreateDir(env, "")` calls itself with the identical empty argument forever.

Confirmed with gdb: about 29,100 identical recursive frames before the stack overflows and the process crashes. Call chain: `DebugIdentityOp::Compute` (and the sibling `DebugIdentityV3Op`, `DebugNanCountOp`, `DebugNumericSummaryOp`, all through the shared `BaseDebugOp::PublishTensor`) -> `DebugIO::PublishDebugTensor` -> `DebugFileIO::DumpTensorToEventFile` -> `DumpEventProtoToFile` -> `RecursiveCreateDir` (self-recursive).

## Fix

Add a base case: when the directory is empty, there is no parent left to create, so return immediately instead of recursing. The caller's own `CreateDir` call then proceeds normally, so relative debug dump paths still work, anchored at the current working directory instead of hitting an unreachable "parent."

```cpp
if (dir.empty()) {
  // io::Dirname() returns "" for any path with no '/' component; this is
  // a fixed point (Dirname("") == ""). It means there is no parent
  // directory left to create: any remaining path is relative to the
  // current working directory. Stop here instead of recursing forever.
  return absl::OkStatus();
}
```

Scope: only `RecursiveCreateDir` in `debug_io_utils.cc`. `io::Dirname`/`SplitPath` (in the TSL/XLA-mirrored `path.cc`) are untouched; their empty-string-on-no-separator behavior is relied on elsewhere and is correct as is. There's a separate, pre-existing TOCTOU race noted in a comment a few lines below this fix (the `CreateDir`-then-recheck pattern) that this PR does not address; it's a different issue from the one reported here.

## Testing

Added `DumpTensorToDirWithRelativeDumpRootSunnyDay` to `DebugIOUtilsTest` in `debug_io_utils_test.cc`, using a relative, slash-less dump root to exercise the exact path that used to crash, and asserting the dump succeeds and reads back correctly.

`bazel test //tensorflow/core/debug:debug_io_utils_test`: 15/15 tests pass, including the new one and all pre-existing cases (no regressions).

### CODE REVIEW

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
### Review of Fix for Stack Overflow in `DebugFileIO::RecursiveCreateDir`

The fix correctly addresses the infinite recursion bug by adding a base case for empty directories returned by `io::Dirname()`.

#### Recommendations for regression testing

The added test `DumpTensorToDirWithRelativeDumpRootSunnyDay` uses a relative path. In some environments, the current working directory might not be writable, or running tests with relative paths can leave artifacts or interfere with other tests if the working directory is changed without restoration.

To ensure the test passes reliably across different environments, it is recommended to manage the working directory safely. Directly calling `chdir()` can affect subsequent tests in the same process. A scoped wrapper is recommended to save and restore the working directory.

Example implementation of a scoped directory changer:

```cpp
#include <unistd.h>
#include <string>
#include <cstdlib>

class ScopedChdir {
 public:
  explicit ScopedChdir(const std::string& path) {
    char* cwd = getcwd(nullptr, 0);
    if (cwd) {
      original_cwd_ = cwd;
      free(cwd);
    }
    chdir(path.c_str());
  }
  ~ScopedChdir() {
    if (!original_cwd_.empty()) {
      chdir(original_cwd_.c_str());
    }
  }
 private:
  std::string original_cwd_;
};
```

You can update the test to use this wrapper to safely execute the relative path slice:

```cpp
TEST_F(DebugIOUtilsTest, DumpTensorToDirWithRelativeDumpRootSunnyDay) {
  Initialize();
  // Safely switch to temporary directory for the duration of this test
  ScopedChdir scoped_chdir(testing::TmpDir());

  // A relative directory with no path separator at all.
  const std::string test_dir = "tfdbg_relative_dump_root_test";
  
  // ... rest of the test body ...
}
```

This approach ensures the test is independent and leaves no side effects on the environment.

**Comment by somtri** [comment]
@dmiltr3 Good catch. The test now runs inside a scoped working-directory wrapper, so it doesn't depend on the runner's ambient CWD.

One change from your suggested code: `<unistd.h>`/`chdir()` is POSIX-only and would break TensorFlow's Windows build, and there's no existing `chdir` usage anywhere in `tensorflow/core`'s test suite to defer to. I guarded the wrapper with `PLATFORM_WINDOWS` instead, using `_chdir`/`_getcwd` from `<direct.h>` on Windows and `chdir`/`getcwd` from `<unistd.h>` elsewhere. This matches the `PLATFORM_WINDOWS` guard pattern already used in `tensorflow/core/platform/env_test.cc`. The wrapper chdir's into `testing::TmpDir()`, which is writable in every CI environment.

This doesn't weaken what the test is checking. `io::Dirname()` is a pure string operation and never touches the filesystem, so the bug (`Dirname("")` returning `""` for a bare relative name) still reproduces the same way regardless of CWD. I confirmed this directly rather than just assuming it: I temporarily reverted the `dir.empty()` fix, reran the test with the chdir wrapper in place, and it hung with a stack overflow exactly as before. Then I restored the fix and reran the full suite (15/15 pass).

Pushed in the latest commit.

**Review by dmiltr3 — APPROVED** [review]
Thank you for addressing the review comments! The base case in `RecursiveCreateDir` for empty paths (`io::Dirname()` returning `""`) fixes the infinite recursion bug on relative paths, and the cross-platform `ScopedChdir` test wrapper cleanly isolates the working directory during unit testing. Looks great to merge.

**Comment by somtri** [comment]
Rebased onto current master to get a fresh CI run.

The earlier `build-windows-x86` red was unrelated to this change. It failed linking `_pywrap_tensorflow_common.dll` with `undefined symbol: ?SetOptions@ConverterPassOptionsSetter@TFL@mlir@@UEBAXAEAULargeConstantFoldPassOptions@23@@Z`, and that run reported 952 test failures. That link error is gone on current master: #124966 built Windows clean there earlier today.

`import/copybara` has reported "An error happened while migrating the change" since 2026-08-06. Is that visible on your side? The new run is currently gated at `action_required`.

**Comment by somtri** [comment]
Ping on this one. It was approved on Aug 6 and had `ready to pull`, but I rebased onto master on Aug 10 to get a clean CI run past an unrelated `build-windows-x86` link failure, and the bot dropped the label two seconds after the push.

The rebase changed no code. GitHub still shows the PR as approved against the current head (e829051), and CI, the license check and the Actions scan all pass on it.

@dmiltr3 could you re-apply `ready to pull` when you get a chance?

**Review by dmiltr3 — APPROVED** [review]
The changes in this pull request correctly address the infinite recursion in `DebugFileIO::RecursiveCreateDir` for relative paths, all CI checks pass, and internal presubmit tests have passed cleanly. Ready for merge.
